dogsdogsdogs/src/lib.rs

```diff
@@ -343,21 +358,21 @@ where
                                 cursor.seek_key(&storage, &key);
                                 if cursor.get_key(&storage) == Some(&key) {
                                     while let Some(value) = cursor.get_val(&storage) {
-                                        let mut count = 0;
+                                        let mut count = R::zero();
                                         cursor.map_times(&storage, |t, d| if t.less_equal(time) { count += d; });
                                         // assert!(count >= 0);
-                                        if count > 0 {
+                                        if count > R::zero() {
```

Comment:
Can you talk me through this logic? I would have though that `Monoid` did not imply `Ord`, and that our goal here is probably to *multiply* `count` and `diff`. Is that wrong?

Reply:
Same thing over in `validate()`; I think our goal (correct me if wrong) is to end up multiplying all of the constraints together. Back in relational-land this would implement "intersect", and in general monoid land it implements something else, I suppose...

Reply:
Ah I must have missed pushing the multiplication part.
As far as I understand it, the 
```rust
cursor.map_times(&storage, |t, d| if t.less_equal(time) { count += d; });
```
both in `validate` and `propose` starts with the zero element of the monoid and then applies its addition for every time in the cursor, sort of consolidating one specific (k, v)-pair in the `propose_trace`/`validate_trace`.
If its non-zero we create an output like so:
```rust
if count > R::zero() {
    session.give(((prefix.clone(), value.clone()), time.clone(), diff.clone() * count));
}
```
using the `Mul` that the semi-ring supplies.
Same for validate.

Reply:
I suspect rather than `> R::zero()` you'll probably want `!count.is_zero()`, and you may even want to perform the multiplication with `diff` first (to see if they annihilate).

I think the `>` you have access to is not through `Monoid`, but rather just the `Ord` implementation that differential requires to sort things. I don't think it is semantically meaningful here (e.g., the numbers could accumulate negatively, and you would want to multiply anyhow).

Reply:
And, if you have the free cycles, there are several other spots in the code where you have `!= R::zero()` that we should probably make into `!blah.is_zero()`, in potential anticipation of going from monoids to semigroups (where a zero element may not inhabit the type). Not as urgent, but it was something that came to mind as I read the edits.

Reply:
Mh yes good point. Didn't think much about the `>`.
Let me think a bit about semi-rings, but multiplication first wouldn't cost us anything, so seems like a sensible choice.
I am not sure if we can have two non-zero elements of our semiring using `Mul` would turn into the identiy, aka zero, element of `Add`

Reply:
Sorry, replying to the github email apparently put the response one level up. I'll avoid using that mechanism in the future.

Reply:
Oh yeah didn't think about more complex diffs then a single one^^
True that, true that